### ISSUE TITLE
ST06 + LT04 corrupts SQL (commas are erroneously added or removed)

### ISSUE THREAD

**Issue by daviewales** [sqlfluff/sqlfluff#4882]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have the following sqlfluff config:

``` cfg
[sqlfluff]
dialect = tsql
max_line_length = 88

[sqlfluff:indentation]
allow_implicit_indents = True

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True
additional_allowed_characters = ['-']
```

Given the following SQL as input, running `sqlfluff fix` will add or remove commas when it shouldn't, rendering the SQL invalid:

``` sql
with cte1 as (
select "a"
      ,"b"
      ,coalesce("g1"
               ,"g2"
               ,"g3"
       ) as "g_combined"

      ,"i"
      ,"j"
  from test
),

cte2 as (
select "col1"

      ,'start: ' + "col2" as "new_col2"

      ,'start2: ' + "col3" as "new_col3"

      ,"col4"
      ,"col5"
from cte1
),

select * from cte2
```

### Expected Behaviour

I expect the commas to be correctly placed.

### Observed Behaviour

``` sql
with cte1 as (
    select
        a,
        b,
        i    -- <-- Missing comma

        j,
        coalesce(
            g1,
            g2,
            g3
        ) as g_combined,
    from test
),

cte2 as (
    select
        col1,

        col4    -- <-- Missing comma

        col5    -- <-- Missing comma

        'start: ' + col2 as new_col2,,        -- <-- Extra comma
        'start2: ' + col3 as new_col3,
    from opening_hours_open_close_cols
),

select * from cte2
```

### How to reproduce

Use the `.sqlfluff` config and the SQL file included above. Run `sqlfluff fix`.

### Dialect

tsql

### Version

sqlfluff, version 2.1.0
Python 3.10.10

### Configuration

[sqlfluff]
dialect = tsql
max_line_length = 88

[sqlfluff:indentation]
allow_implicit_indents = True

[sqlfluff:rules:references.special_chars]
allow_space_in_identifier = True
additional_allowed_characters = ['-']

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by daviewales** [comment]
Seems related but different to #1734
